### PR TITLE
Expand 'yield' in internal macro calls.

### LINKED ISSUE #27: yield inside tokio select! doesn't work
Yielding [here](https://github.com/tekjar/rumq/blob/master/rumq-client/src/eventloop.rs#L123) inside tokio `select!` doesn't work. I can bypass the error with a variable but I'm facing a compiler crash with assignments like this
```
exit = Notification::StreamEnd(e.into());
```
So I'm mixing yields and assignments to make the code compile.

As the description says, yielding inside `select!` will be helpful for my use case :)

### PR DESCRIPTION
**Pull request by SergioBenitez**
Resolves #27.

As in the title.

### CODE REVIEW

**Comment by SergioBenitez** [comment]
I can't replicate the test failures locally. It seems the test CI is simply running `cargo test +stable --all-features`, which passes on my machine with flying colors. What am I missing?

File: async-stream/tests/ui/yield_in_async.stderr [hunk 1]
```diff
@@ -24,6 +24,11 @@ error[E0271]: type mismatch resolving `<[static generator@$DIR/src/lib.rs:202:9:
 10 | |     };
    | |______^ expected `()`, found integer
    |
+  ::: $RUST/core/src/future/mod.rs
+   |
+   |       T: Generator<ResumeTy, Yield = ()>,
+   |                              ---------- required by this bound in `from_generator`
+   |
```
Comment (taiki-e) [inline async-stream/tests/ui/yield_in_async.stderr:31]:
```suggestion
```

This appears when running locally, but not in CI.

**Review by taiki-e — COMMENTED** [review]
> I can't replicate the test failures locally. It seems the test CI is simply running `cargo test +stable --all-features`, which passes on my machine with flying colors. What am I missing?

I'm not sure if it's a trybuild bug or a rustc bug, (sometimes) test results in CI may differ from those run locally...

**Review by taiki-e — COMMENTED** [review]
Thanks! This looks good to me aside from a few nits.

File: async-stream-impl/src/lib.rs [hunk 2]
```diff
@@ -109,9 +177,31 @@ impl VisitMut for Scrub<'_> {
         }
     }
 
-    fn visit_item_mut(&mut self, _: &mut syn::Item) {
-        // Don't transform inner items.
+    fn visit_macro_mut(&mut self, mac: &mut syn::Macro) {
+        let segments = &mac.path.segments;
+        if segments.last().map_or(false, |s| s.ident == "stream") {
```
[the remark this answers is not in the document]
Comment (taiki-e) [reply async-stream-impl/src/lib.rs:182]: (It doesn't have to be part of this PR, but it seems we need to have tests like try_stream_in_stream, try_stream_in_try_stream, stream_in_try_stream anyway.)

**Comment by SergioBenitez** [comment]
Cool. This looks good to me, too!

**Review by taiki-e — APPROVED** [review]

**Comment by SergioBenitez** [comment]
What's keeping this from being merged?

**Comment by taiki-e** [comment]
Sorry, I forgot about this, I'll review it again and merge it if it looks good.

File: async-stream-impl/src/lib.rs [hunk 3]
```diff
@@ -34,6 +34,68 @@ impl<'a> Scrub<'a> {
...
+                        *modified = true;
+                        tokens = rest.into_iter();
+                    }
+                    Err(e) => {
+                        out.append_all(&mut e.to_compile_error().into_iter());
+                        *modified = true;
+                        return;
+                    }
+                }
+            }
```
Comment (taiki-e) [inline async-stream-impl/src/lib.rs:73]: I'd like to add the branch like the following and skip visits to the other stream macro.

```suggestion
            }
            TokenTree::Ident(i) if i == "stream" || i == "try_stream" => {
                out.append(TokenTree::Ident(i));
                match tokens.peek() {
                    Some(TokenTree::Punct(p)) if p.as_char() == '!' => {
                        out.append(tokens.next().unwrap()); // !
                        out.append(tokens.next().unwrap()); // assuming { .. } or [ .. ] or ( .. )
                    }
                    _ => {}
                }
            }
```

**Review by taiki-e — COMMENTED** [review]
I noticed this breaks code that has another async stream macro inside the macro.

```rust
use tokio::select;
use async_stream::{stream, try_stream};
use futures_util::stream::StreamExt;

async fn do_stuff_async() {}

let s = stream! {
    select! {
        _ = do_stuff_async() => {
            let another_s = try_stream! {
                yield;
            };
            let _: Result<(), ()> = Box::pin(another_s).next().await.unwrap();
        },
        else => {},
    }
    yield
};
```

**Comment by SergioBenitez** [comment]
Good catch! Should be good to go.

**Comment by SergioBenitez** [comment]
Fixed the rustfmt issue.

**Review by taiki-e — APPROVED** [review]
Thanks!

**Comment by SergioBenitez** [comment]
Any chance this can find its way into a release? I'm currently exporting [`stream`](https://api.rocket.rs/master/rocket/response/stream/macro.stream.html) from my fork but would love to depend on upstream instead. This would also make it possible to publish a release that depends on the macro.

**Comment by taiki-e** [comment]
Filed https://github.com/tokio-rs/async-stream/pull/60 for next release.

**Comment by taiki-e** [comment]
Published in v0.3.2.

**Comment by SergioBenitez** [comment]
Awesome, awesome, awesome. Thank you!

**Pull request merged**